crates/ruff_python_formatter/src/comments/visitor.rs

```diff
@@ -0,0 +1,354 @@
+            walk_expr(self, expr);
+        }
+        self.finish_node(expr);
+    }
+
+    fn visit_comprehension(&mut self, comprehension: &'ast Comprehension) {
+        if self.start_node(comprehension).is_traverse() {
+            walk_comprehension(self, comprehension);
+        }
+        self.finish_node(comprehension);
+    }
+
+    fn visit_excepthandler(&mut self, excepthandler: &'ast Excepthandler) {
+        if self.start_node(excepthandler).is_traverse() {
+            walk_excepthandler(self, excepthandler);
+        }
+        self.finish_node(excepthandler);
+    }
+
+    fn visit_format_spec(&mut self, format_spec: &'ast Expr) {
+        if self.start_node(format_spec).is_traverse() {
+            walk_expr(self, format_spec);
+        }
+        self.finish_node(format_spec);
+    }
+
+    fn visit_arguments(&mut self, arguments: &'ast Arguments) {
+        if self.start_node(arguments).is_traverse() {
+            walk_arguments(self, arguments);
+        }
+        self.finish_node(arguments);
+    }
+
+    fn visit_arg(&mut self, arg: &'ast Arg) {
+        if self.start_node(arg).is_traverse() {
+            walk_arg(self, arg);
+        }
+        self.finish_node(arg);
+    }
+
+    fn visit_keyword(&mut self, keyword: &'ast Keyword) {
+        if self.start_node(keyword).is_traverse() {
+            walk_keyword(self, keyword);
+        }
+        self.finish_node(keyword);
+    }
+
+    fn visit_alias(&mut self, alias: &'ast Alias) {
+        if self.start_node(alias).is_traverse() {
+            walk_alias(self, alias);
+        }
+        self.finish_node(alias);
+    }
+
+    fn visit_withitem(&mut self, withitem: &'ast Withitem) {
+        if self.start_node(withitem).is_traverse() {
+            walk_withitem(self, withitem);
+        }
+
+        self.finish_node(withitem);
+    }
+    fn visit_match_case(&mut self, match_case: &'ast MatchCase) {
+        if self.start_node(match_case).is_traverse() {
+            walk_match_case(self, match_case);
+        }
+        self.finish_node(match_case);
+    }
+
+    fn visit_pattern(&mut self, pattern: &'ast Pattern) {
+        if self.start_node(pattern).is_traverse() {
+            walk_pattern(self, pattern);
+        }
+        self.finish_node(pattern);
+    }
+}
+
+fn text_position(comment_range: TextRange, source_code: SourceCode) -> CommentTextPosition {
+    let before = &source_code.as_str()[TextRange::up_to(comment_range.start())];
+
+    for c in before.chars().rev() {
+        match c {
+            '\n' | '\r' => {
+                break;
+            }
+            c if is_python_whitespace(c) => continue,
+            _ => return CommentTextPosition::EndOfLine,
+        }
+    }
+
+    CommentTextPosition::OwnLine
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum TraversalSignal {
+    Traverse,
+    Skip,
+}
+
+impl TraversalSignal {
+    const fn is_traverse(self) -> bool {
+        matches!(self, TraversalSignal::Traverse)
```

Comment:
If you want, you can derive `is_macro::Is` for this.

Reply:
I've mixed feelings on `is_macro::Is` ;)